ipa-core/src/net/test.rs

```diff
@@ -10,33 +10,127 @@
     hpke::{Deserializable as _, IpaPublicKey},
     net::{ClientIdentity, Helper, MpcHelperClient, MpcHelperServer},
+    sharding::ShardIndex,
     sync::Arc,
     test_fixture::metrics::MetricsHandle,
 };
 
-pub const DEFAULT_TEST_PORTS: [u16; 3] = [3000, 3001, 3002];
+/// Simple struct to keep default port configuration organized.
+pub struct Ports<C> {
+    ring: C,
+    shards: C,
+}
+
+/// A **single** ring with 3 hosts, each with a ring and sharding port.
+pub const DEFAULT_TEST_PORTS: Ports<[u16; 3]> = Ports {
+    ring: [3000, 3001, 3002],
+    shards: [6000, 6001, 6002],
+};
+
+/// Configuration of a server that can be reached via socket or port.
+pub struct AddressableTestServer {
+    /// Contains the ports
+    pub config: ServerConfig,
+    /// Sockets are created if no port was specified.
+    pub socket: Option<TcpListener>,
+}
+
+/// A bunch of addressable servers. These can be part of a Shard or MPC network.
+#[derive(Default)]
+pub struct TestServers {
+    pub configs: Vec<AddressableTestServer>,
+}
+
+impl TestServers {
+    /// Creates a bunch of addressable servers from the given configurations plus the optional
+    /// ports. Meant to be used during the definition of an MPC ring.
+    fn new(configs: Vec<ServerConfig>, sockets: Vec<Option<TcpListener>>) -> Self {
+        assert_eq!(configs.len(), sockets.len());
+        let mut new_configs = Vec::with_capacity(configs.len());
+        for s in zip(configs, sockets) {
+            new_configs.push(AddressableTestServer {
+                config: s.0,
+                socket: s.1,
+            });
+        }
+        TestServers {
+            configs: new_configs,
+        }
+    }
+
+    /// Adds a test server. These increments are useful during shard network definition.
+    fn push_shard(&mut self, config: AddressableTestServer) {
+        self.configs.push(config);
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &AddressableTestServer> {
+        self.configs.iter()
+    }
+}
 
+/// Either a single Ring on MPC connection or all of the shards in a Helper.
+pub struct TestNetwork<F: ConnectionFlavor> {
+    pub network: NetworkConfig<F>, // Contains Clients config
+    pub servers: TestServers,
+}
+
+impl TestNetwork<Shard> {
+    #[must_use]
+
+    /// Gets a ref to the first shard in this network.
+    pub fn get_first_shard(&self) -> &AddressableTestServer {
+        self.servers.configs.first().unwrap()
+    }
+
+    /// Gets a mut ref to the first shard in this network.
+    pub fn get_first_shard_mut(&mut self) -> &mut AddressableTestServer {
+        self.servers.configs.get_mut(0).unwrap()
+    }
+}
+
+/// Uber container for test configuration. Provides access to a vec of MPC rings and 3 sharding
+/// networks (one for each Helper)
 pub struct TestConfig {
     pub disable_https: bool,
-    pub network: NetworkConfig<Helper>,
-    pub servers: [ServerConfig; 3],
-    pub sockets: Option<[TcpListener; 3]>,
+    pub rings: Vec<TestNetwork<Helper>>,
```

Comment:
nit: I think we need consistent naming for variables here. `rings`/`shards` or something along these lines